Tests/Unit/LoRaTools/MacCommandTests.cs

```diff
@@ -0,0 +1,293 @@
+
+        [Fact]
+        public void Cid_Success() => Assert.Equal(Cid, Subject.Cid);
+
+        protected void DeserializationTest(Func<T, object> transform, string json) =>
+            Assert.Equal(transform(Subject), transform(JsonConvert.DeserializeObject<T>(JsonUtil.Strictify(json)) ?? throw new InvalidOperationException("JSON was deserialized to null.")));
+
+        protected void FromBytesTest(Func<T, object> transform, Func<byte[], T> subject) =>
+            Assert.Equal(transform(Subject), transform(subject(GetFullBytes(DataBytes))));
+
+        private byte[] GetFullBytes(IReadOnlyList<byte> dataBytes) => dataBytes.Prepend((byte)Cid).ToArray();
+    }
+
+    public sealed class TxParamSetupRequestTests : MacCommandTests<TxParamSetupRequest>
+    {
+        public TxParamSetupRequestTests() :
+            base(Cid.TxParamSetupCmd,
+                 new TxParamSetupRequest(new DwellTimeSetting(false, false, 0)),
+                 new byte[] { 0b0000_0000 })
+        { }
+
+        [Fact]
+        public void Init_Throws_When_Dwell_Time_Settings_Are_Empty() =>
+            Assert.Throws<ArgumentNullException>(() => new TxParamSetupRequest(null!));
+
+        [Fact]
+        public void Init_Throws_When_Eirp_Is_Invalid() =>
+            Assert.Throws<ArgumentException>(() => new TxParamSetupRequest(new DwellTimeSetting(false, false, 16)));
+
+        public static TheoryData<DwellTimeSetting, byte> ToBytes_Theory_Data() => TheoryDataFactory.From(new (DwellTimeSetting, byte)[]
+        {
+            (new DwellTimeSetting(false, false, 0), 0b0000_0000),
+            (new DwellTimeSetting(true, false, 0), 0b0010_0000),
+            (new DwellTimeSetting(false, true, 0), 0b0001_0000),
+            (new DwellTimeSetting(false, false, 13), 0b0000_1101),
+            (new DwellTimeSetting(true, true, 15), 0b0011_1111),
+        });
+
+        [Theory]
+        [MemberData(nameof(ToBytes_Theory_Data))]
+        public void ToByte_Success_Cases(DwellTimeSetting dwellTimeSetting, byte expected) =>
+            ToBytes_Internal(new TxParamSetupRequest(dwellTimeSetting), new[] { expected });
+    }
+
+    public sealed class TxParamSetupAnswerTests : MacCommandTests<TxParamSetupAnswer>
+    {
+        public TxParamSetupAnswerTests() :
+            base(Cid.TxParamSetupCmd,
+                 new TxParamSetupAnswer(),
+                 Array.Empty<byte>())
+        { }
+    }
+
+    public sealed class DevStatusAnswerTests : MacCommandTests<DevStatusAnswer>
+    {
+        public DevStatusAnswerTests() :
+            base(Cid.DevStatusCmd,
+                 new DevStatusAnswer(1, 2),
+                 new byte[] { 1, 2 })
+        { }
+
+        [Fact]
+        public void FromBytes_Success() => FromBytesTest(command => new { command.Cid, command.Battery, command.Margin },
+                                                         bytes => new DevStatusAnswer(new ReadOnlySpan<byte>(bytes)));
```

Comment:
nit: you can use `AsSpan()` here and elsewhere for similar cases:

```suggestion
                                                         bytes => new DevStatusAnswer(bytes.AsSpan()));
```

Reply:
That's indeed better, done in 9b13c9464b2059b1e6da20fc58dce7586574bc31